packages/backend/src/models/ProjectModel/ProjectModel.ts

```diff
@@ -747,6 +752,335 @@ export class ProjectModel {
         }
     }
 
+    async duplicateContent(projectUuid: string, previewProjectUuid: string) {
+        Logger.debug(
+            `Duplicating content from ${projectUuid} to ${previewProjectUuid}`,
+        );
+
+        return this.database.transaction(async (trx) => {
+            const previewProject = await trx('projects')
+                .where('project_uuid', previewProjectUuid)
+                .first();
+            const spaces = await trx('spaces')
+                .leftJoin(
+                    'projects',
+                    'projects.project_id',
+                    'spaces.project_id',
+                )
+                .where('projects.project_uuid', projectUuid)
+                .select('spaces.*');
+
+            Logger.debug(
+                `Duplicating ${spaces.length} spaces on ${previewProjectUuid}`,
+            );
+            const spaceIds = spaces.map((s) => s.space_id);
+
+            const newSpaces =
+                spaces.length > 0
+                    ? await trx('spaces')
+                          .insert(
+                              spaces.map((d) => ({
+                                  ...d,
+                                  space_id: undefined,
+                                  space_uuid: undefined,
+                                  project_id: previewProject?.project_id,
+                              })),
+                          )
+                          .returning('*')
+                    : [];
+
+            const spaceMapping = spaces.map((s, i) => ({
+                id: s.space_id,
+                newId: newSpaces[i].space_id,
+            }));
+
+            const getNewSpace = (oldSpaceId: number): number =>
+                spaceMapping.find((s) => s.id === oldSpaceId)?.newId!;
+            const spaceShares = await trx('space_share').whereIn(
+                'space_id',
+                spaceIds,
+            );
+
+            const newSpaceShare =
+                spaceShares.length > 0
+                    ? await trx('space_share')
+                          .insert(
+                              spaceShares.map((d) => ({
+                                  ...d,
+                                  space_id: getNewSpace(d.space_id),
+                              })),
+                          )
+                          .returning('*')
+                    : [];
+
+            const charts = await trx('saved_queries')
+                .leftJoin('spaces', 'saved_queries.space_id', 'spaces.space_id')
+                .leftJoin(
+                    'projects',
+                    'projects.project_id',
+                    'spaces.project_id',
+                )
+                .where('projects.project_uuid', projectUuid)
+                .select('saved_queries.*');
```

Comment:
For example here, would we want to type `charts`? Or are we flexible with that in this case?

Reply:
Understood, also GH copilot isn't smart enough to predict the types for us... 😆 it's a lot of work